Replace Amplify authentication with anonymous-only login

ktohalloran opened the pull request:
## Overview

This PR completely removes Amplify and all of its related frontend authentication components and replaces the login flow with anonymous log in by replacing the `Authentication` HOC with a smaller `anonymousAuthenticator` HOC.

### Demo

![login_AdobeCreativeCloudExpress](https://user-images.githubusercontent.com/77936689/162217628-4b89bb31-68aa-49dc-8fdb-1ac479beb320.gif)

You are now only able to log in anonymously; doing so will take you directly to fill out your profile information.

### Notes

I had a tough time deciding the extend of what to cut out versus leave in for the sake of building from for Django authentication. Ultimately, along with removing anything that was explicitly related to Amplify's authentication process, I chose to remove places where currently useless checks for anonymity and/or authentication level impacted code readability prior to the addition of Django authentication, but left them in where they could be helpful for future authentication but not distracting. An example of this is removing the checks and related code for `isAnonymous` and `isCounselor` in `edit-profile.js`, as an anonymous user should never be presented with the option to re-send their invitation email, but leaving `clientAccountConfirmed`, `clientEmail`, and `clientInviteSent` as properties on the `AccountProfile` type. Following this rule, I have removed all checks for `isAnonymous`, `isCounselor`, and account validation, but have left `EMAIL_REGEX` in `constants.js` and email-related strings in the translation files.

One strange behavior I found following these changes was if you clicked "Continue without an account" and then clicked the browser's back button before completing the form, it would take you to a blank "search" page, rather than returning to the login page, unless you cleared your browser's local storage. This was compared to `develop`, where clicking back after selecting anonymous login takes you back to wherever you navigated to the login page from. I fixed this by replacing the redirect ternaries in `application.js` to redirect to '/profile' instead of '/search'. This fix is a little odd in the case of the 3rd path to '/profile', but it seemed a decent placeholder until we restore authentication.

## Testing Instructions

 * Run `./scripts/setup`
 * Run `./scripts/server` and navigate to **http://localhost:9966/**
 * Select "Continue without an account". It should take you to the profile page. 
 * Click your browser's back button and verify that it doesn't take you to a blank search page. ~~Then fill out the form and click save.~~
 * Click Cancel and verify that it takes you back to the login page. Select "Continue without an account" again and fill out the form and click save.
 * Verify that you can favorite neighborhoods, make changes to your profile, and that your profile is saved on browser refresh.

EDIT: Added testing instructions for re-route to login on cancel during profile edit.


Resolves #431

Commit:
Replace login and edit profile with anonymous-only authentication

This commit replaces the Amplify components from custom-sign-in,
index, edit-profile, and custom-header-bar with an anonymous-only
authentication wrapper. It also removes reference to checks for
isCounselor and isAnonymous, as well as other parts of client
profiles (such as email validation and profile confirmation) that
are not relevant when the client is anonymous.

Commit:
Replace references to Amplify's changeUserProfile

This replaces Amplify's changeUserProfile method with the new
anonymous authenticator's handleProfileChange method in the dock,
main page, and neighborhood detail pages.

Commit:
Restore saving profile to local storage for multiple sessions

This commit prevents the account from completely resetting on
refresh or closing the browser.

Commit:
Update styling for login page to replace amplify-theme

Commit:
Remove Amplify as dependency and all frontend references to it

Commit:
Replace redirect to '/search' page with redirect to '/profile'

BryanQuigley commented on taui/src/components/anonymous-authenticator.js:
```diff
@@ -0,0 +1,72 @@
+      super(props)
+
+      this.handleAnonymousLogin = this.handleAnonymousLogin.bind(this)
+      this.handleProfileChange = this.handleProfileChange.bind(this)
```
I somewhat expected handleProfileChange to do the LoadProfile().  Any reason it doesn't?

ktohalloran replied:
I use `loadProfile()` as passed as a prop from `Application` on line 20--did you expect to see it used in a different way or as a method of `anonymousAuthenticator` rather than passed as a prop?

BryanQuigley replied:
Conceptually I just expect handleProfileChange to "handle" loading the profile. It's fine as is, I was just curious.

ktohalloran replied:
Ah, I see. That's a good point--I had made `handleProfileChange` to be a counterpart to Amplify's `withAuthenticator.changeUserProfile` which also doesn't use `loadProfile`, which makes sense within Amplify's structure but maybe not within this less intense `anonymousAuthenticator`. I'll explore that structure a bit more--thanks for pointing that out!

BryanQuigley reviewed (approved):
The only small improvement I'd suggest is to just redirect back to the home page if someone presses Cancel and they haven't set up a profile yet.

ktohalloran commented:
Thanks for reviewing @BryanQuigley! I agree that would definitely make the user experience better; I'll see what I can do. I think that might be a bit more complicated than I expect just because as soon as you click that "Continue without an account", an empty profile is stored in local storage; anytime you try to go to the login page from there, it currently re-routes to keep you within the app on seeing that you already have that profile. I'll see if I can find a better way to handle that than just sending you right back to "/profile", though.

ddohler commented:
Reassigning to @rachelekm since I'll be out next week.

Commit:
Redirect to login on edit profile cancel before form completion

Initially, once users selected "Continue without an account", they
would be taken to the edit profile page with no ability to go back.
This commit allows users who have not finished filling out the profile
form to return to the login page.

ktohalloran commented:
I've made the edit to the automatic re-routing so that pressing cancel before completing the edit profile form takes you back to login instead of the same blank profile page and have added tested instructions above to check that; however, when I tried moving `loadProfile` into `anonymousAuthenticator`'s `handleProfileChange` method, I got strange behavior on refresh as the app re-rendered without triggering `handleProfileChange`. The `loadProfile` method seems to be ok where it is only activating on app render and with local storage updating via `storeConfig`, so I'm going to leave it as is.

rachelekm commented on taui/src/components/anonymous-authenticator.js:
```diff
@@ -0,0 +1,72 @@
+import CustomSignIn from './custom-sign-in'
+import CustomHeaderBar from './custom-header-bar'
+
+export default function anonymousAuthenticator (Comp) {
```
Do we anticipate updating state here for any tokens/credentials when we shift to Django? Would it be better to rename to just `Authenticator` to keep its purpose open or is this just going to be used only for anonymous users?

rachelekm commented on taui/src/components/custom-sign-in.js:
```diff
@@ -42,80 +24,23 @@ class SignInHeader extends React.Component {
-        </FormSection>
+        <div className='auth-screen__main auth-main'>
+          <div className='auth-main__anonymous-login'>
+            {t('SignIn.AnonymousExplanation') + ' '}
```
```suggestion
            {`${t('SignIn.AnonymousExplanation') }`}
```
This doesn't really matter but I think the general pattern in this repo is to use template literals

rachelekm commented on taui/src/components/anonymous-authenticator.js:
```diff
@@ -0,0 +1,72 @@
+
+    handleProfileChange (profile: AccountProfile) {
+      storeConfig(PROFILE_CONFIG_KEY, profile)
+      this.props.store.dispatch({type: 'set profile', payload: profile})
```
```suggestion
      this.props.setProfile(profile)
```
This prop calls both and has some error handling

rachelekm commented on taui/src/components/edit-profile.js:
```diff
@@ -248,43 +191,8 @@ class EditProfile extends PureComponent<Props> {
+    this.props.history.push('/map')
   }
 
   getUserDataVoucherNumber (userData: any): string {
```
I think we can get rid of `getUserDataVoucherNumber` now since it's not used by anything

rachelekm reviewed (commented):
Looks great! I left a couple small comments and then have one more broad thought on the `AnonymousAuthenticator` HOC. I see that at this point the user profile is dual-purpose: temporary profile information as well as acting as the “login” credentials for anonymous users. But I feel like that's causing the purposes of `handleProfileChange` and `handleAnonymousLogin` to blur together a little, and I agree that it would be nice to move `this.props.loadProfile()` (although I see your note above so maybe not this one if it's tricky) as well as the default `anonymousProfile` creation into `handleProfileChange` to better organize all profile changes into one function. That would make `handleAnonymousLogin` kind of obsolete, but I think it might be ok to get rid of it because it’s only being called once at this point — and even then it’s just to set the profile so it’s not really functionally “logging in” or “logging out”. Do you know if we’re going to eventually have some kind of token/credential stored in frontend state for user auth? Maybe in that instance it would be good to keep to handle whatever app-level auth state will exist besides the profile? What do you think?

ktohalloran commented:
@rachelekm thanks for reviewing! For your questions about how this will work when we're using Django, the answer is I don't know. I talked with @ddohler about this, asking if this card was about just putting something in place to replace Amplify or building a foundation for the Django authentication system and he said he "was envisioning more the first option, under the assumption that the anonymous login is kind of a weird separate thing from the rest of the auth system anyway, so it hopefully shouldn't interfere too much with building out the rest of the system later... it's expected that we won't know exactly what the Django-based auth will look like right now. I would say to try to generally structure things for easy alteration now, but the main goal here is to get rid of Amplify and put a placeholder there so that we still have site access while we work on the rest of the auth, so it's fine if things need to be adjusted later." So based on that, I'm assuming the point of this card is to make anonymous login work as best we can, assuming we can't fully anticipate what Django will need. To that end, I think leaving it named `AnonymousAuthenticator` makes sense, because that's all it may ever be designed to do. But I'll work on consolidating `handleProfileChange` and `handleAnonymousLogin`--you're right they both don't need to exist and doing so will improve readability and (in theory anyway) replaceability.

Commit:
Streamline anonymousAuthenticator methods and clean up code

This commit consolidates anonymousAuthenticator's handleProfileChange
and handleAnonymousLogin into one handleAuthChange method.

ktohalloran commented:
Ok, I've consolidated `handleProfileChange` and `handleAnonymousLogin` into one method that I've renamed to the more generic `handleAuthChange` and made the few other changes you suggested, so this should be ready for re-review.

rachelekm commented on taui/src/components/custom-header-bar.js:
```diff
@@ -1,58 +1,26 @@
-    return nameFromAttr || user.name || user.username
+    this.setState({
+      userProfile: this.props.userProfile
+    })
```
```suggestion
```
Since we're not using `userProfile` outside of `render()` and it's already being set to `this.props.userProfile` there, I don't think we need this

rachelekm reviewed (commented):
Thanks for clarifying how far to take the anonymous authentication, that makes sense and I agree sticking with the naming seems worth it. I also like the consolidated `handleAuthChange`. To better separate auth and profile info further, does it make sense to call the `setProfile` prop directly instead of `handleAuthChange` when setting favorites or making profile changes once "logged in" or is it better to keep the `handleAuthChange` calls there just so everything's consistent? Beyond that question I saw one small thing I missed earlier, but it looks great!

ktohalloran commented:
@rachelekm that's a good question that I've wavered on throughout this whole process. I can see how it would make sense to just call `setProfile` directly; however, I think it comes down to how which method of organization we think makes the most sense here. Making calls directly to `setProfile` outside of `anonymousAuthenticator` largely makes the authenticator HOC obsolete and feels like it would be worth taking it completely from the current method of using a HOC toward a more functional method. Choosing between object oriented vs. functional method of solving the problem was something I contemplated when starting the issue and I did initially try to solve the problem functionally; however, I ended up back at the HOC because I found it to be easier to mimic what Amplify's authenticator was doing. That being said, that was just what I found easiest and is not to say that wouldn't work or that using a HOC is the better approach. I think the question is whether we feel that this object oriented method should be replaced for the sake of readability/replaceability/performance.

rachelekm commented:
Ah cool, thanks for walking me through the process that's helpful context. I see how a HOC is a better fit for our needs and it makes sense to move forward that way 👍
